src/node.h

```diff
@@ -353,18 +354,94 @@ NODE_EXTERN IsolateData* CreateIsolateData(
     ArrayBufferAllocator* allocator = nullptr);
 NODE_EXTERN void FreeIsolateData(IsolateData* isolate_data);
 
-// TODO(addaleax): Add an official variant using STL containers, and move
-// per-Environment options parsing here.
+struct ThreadId {
+  uint64_t id = static_cast<uint64_t>(-1);
+};
+NODE_EXTERN ThreadId AllocateEnvironmentThreadId();
+
+namespace EnvironmentFlags {
+enum Flags : uint64_t {
+  kNoFlags = 0,
+  // Use the default behaviour for Node.js instances.
+  kDefaultFlags = 1 << 0,
+  // Controls whether this Environment is allowed to affect per-process state
+  // (e.g. cwd, process title, uid, etc.).
+  // This is set when using kDefaultFlags.
+  kOwnsProcessState = 1 << 1,
+  // Set if this Environment instance is associated with the global inspector
+  // handling code (i.e. listening on SIGUSR1).
+  // This is set when using kDefaultFlags.
+  kOwnsInspector = 1 << 2,
+  // Set up a standard Node.js Environment, including starting any libuv handles
+  // and the inspector, as well as anything dependent on e.g. environment
+  // variables or command line flags.
+  // This conflicts with LoadEnvironment().
+  // It is recommended to not set or rely on this flag, and it will be removed.
+  kPrepareForExecution = 1 << 3,
+  // Override the --abort-on-uncaught-exception CLI flag to never be active
+  // for this Environment instance.
+  kDoNotAbortOnUncaughtException = 1 << 4
+};
+}  // namespace EnvironmentFlags
+
+// TODO(addaleax): Maybe move per-Environment options parsing here.
 // Returns nullptr when the Environment cannot be created e.g. there are
 // pending JavaScript exceptions.
+// It is recommended to use the second variant taking a flags argument.
+// kPrepareForExecution is used by default for backwards compatibility,
+// but will be removed and LoadEnvironment() should be used instead.
 NODE_EXTERN Environment* CreateEnvironment(IsolateData* isolate_data,
                                            v8::Local<v8::Context> context,
                                            int argc,
                                            const char* const* argv,
                                            int exec_argc,
                                            const char* const* exec_argv);
+NODE_EXTERN Environment* CreateEnvironment(
+    IsolateData* isolate_data,
+    v8::Local<v8::Context> context,
+    const std::vector<std::string>& args,
+    const std::vector<std::string>& exec_args,
+    EnvironmentFlags::Flags flags = static_cast<EnvironmentFlags::Flags>(
+        EnvironmentFlags::kDefaultFlags |
+        EnvironmentFlags::kPrepareForExecution),
+    ThreadId thread_id = {} /* allocates a thread id automatically */);
+
+struct InspectorParentHandle {
+  virtual ~InspectorParentHandle();
+};
+NODE_EXTERN std::unique_ptr<InspectorParentHandle> GetInspectorParentHandle(
+    Environment* env,
+    ThreadId thread_id,
+    const char* url);
```

Comment:
> Why can't `GetInspectorParentHandle()` take the ID using `Environment::thread_id()` because `env` is already the first argument?

I’ve renamed the arguments and added comments to clarify that the first argument is the *parent* env, whereas the other two arguments refer to the *child* environment.

> What's the use case of `InspectorParentHandle`? 

It makes a second Environment accessible to the inspector as if it were a Worker. I’ve added a comment regarding that.

> For now it looks like you can only get this through `GetInspectorParentHandle()` and then tell `LoadEnvironment()` that this is the parent handle, but then why can't `LoadEnvironment()` simply take a coarse-grained `parent_env` and a `url` instead?

That *might* work, but it seems dangerous to access fields in the parent Environment from a child when they might be running on different threads? Basically the race condition issue you mentioned below/above for the idle notifier case.

Reply:
> That might work, but it seems dangerous to access fields in the parent Environment from a child when they might be running on different threads?

I think the race only happens if one thread is writing a field while the other one is reading it? But how is that possible for users if `env->inspector_agent_` is only modified directly in the Environment constructor/destructor? On the other hand how much more dangerous it is compared to calling `GetInspectorParentHandle()` on different threads? What happens if you call `GetInspectorParentHandle()` from a different thread while the environment is being loaded from another?

Reply:
> > That might work, but it seems dangerous to access fields in the parent Environment from a child when they might be running on different threads?
> 
> I think the race only happens if one thread is writing a field while the other one is reading it? But how is that possible for users if `env->inspector_agent_` is only modified directly in the Environment constructor/destructor? 

I’m more worried about `Agent::GetParentHandle()` relying on state that can be modified while the parent Environment is running, than about the inspector agent changing here.

> On the other hand how much more dangerous it is compared to calling `GetInspectorParentHandle()` on different threads? What happens if you call `GetInspectorParentHandle()` from a different thread while the environment is being loaded from another?

I’m not really worried about this being accidentally misused. I took this approach because it comes with a relatively clear flow of data – one function that returns a unique_ptr, one that takes it as an argument.

I’ve added a comment about where this method should be called, for clarity, but generally it fits in with all the other methods (namely that Node.js APIs generally aren’t thread-safe).

Reply:
> I’m more worried about Agent::GetParentHandle() relying on state that can be modified while the parent Environment is running, than about the inspector agent changing here.

In that case, passing the parent handles around is not any less racy than passing the parent environment around? I think ideally one would want to keep `Agent::GetParentHandle()` as close to `Agent::SetParentHandle` and `Agent::Start` as possible (there are no any other places in `Environment` where the parent handle is used I believe). If this method is not exposed, we might still be able to make it less racy with locks somehow, otherwise the caller gets extra responsibility of not misusing it, without much to trade for.